core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala

```diff
@@ -482,94 +455,72 @@ class TiBatchWriteTable(
           buf.toList
         }
 
-        def genNextHandleBatch(
-            batch: List[WrappedRow]): (util.List[Array[Byte]], util.List[Handle]) = {
-          val list = new util.ArrayList[Array[Byte]]()
-          val handles = new util.ArrayList[Handle]()
-          batch.foreach { wrappedRow =>
-            val bytes = buildRowKey(wrappedRow.row, wrappedRow.handle).bytes
-            list.add(bytes)
-            handles.add(wrappedRow.handle)
-          }
-          (list, handles)
-        }
-
         def genNextUniqueIndexBatch(
             batch: List[WrappedRow],
-            index: TiIndexInfo): (util.List[Array[Byte]], util.List[TiRow]) = {
+            index: TiIndexInfo): util.List[Array[Byte]] = {
           val keyList = new util.ArrayList[Array[Byte]]()
-          val rowList = new util.ArrayList[TiRow]()
           batch.foreach { wrappedRow =>
             val encodeResult = buildUniqueIndexKey(wrappedRow.row, wrappedRow.handle, index)
             if (!encodeResult._2) {
               // only add the key if handle is not appended, since if handle is appened,
               // the value must be a new value
-              val bytes = encodeResult._1.bytes
-              keyList.add(bytes)
-              rowList.add(wrappedRow.row)
+              keyList.add(encodeResult._1.bytes)
             }
           }
-          (keyList, rowList)
+          keyList
         }
 
-        def decodeHandle(row: Array[Byte]): Handle = {
-          RowKey.decode(row).getHandle
-        }
-
-        def processHandleDelete(
-            oldValueList: java.util.List[BytePairWrapper],
-            handleList: java.util.List[Handle]): Unit = {
+        def processHandleDelete(oldValueList: java.util.List[BytePairWrapper]): Unit = {
           for (i <- 0 until oldValueList.size) {
             val oldValuePair = oldValueList.get(i)
             val oldValue = oldValuePair.getValue
-            val handle = handleList.get(i)
+            val oldKey = oldValuePair.getKey
 
             if (oldValue.nonEmpty && !isNullUniqueIndexValue(oldValue)) {
-              val oldRow = TableCodec.decodeRow(oldValue, handle, tiTableInfo)
-              rowBuf += WrappedRow(oldRow, handle)
+              val oldHandle = RowKey.decode(oldKey).getHandle
+              val oldRow = TableCodec.decodeRow(oldValue, oldHandle, tiTableInfo)
+              rowBuf += WrappedRow(oldRow, oldHandle)
             }
           }
         }
 
         def processNextBatch(): Unit = {
           rowBuf = mutable.ListBuffer.empty[WrappedRow]
-
+          val mayConflictRowKeys = mutable.Set.empty[Array[Byte]]
```

Comment:
why need mayConflictRowKeys, and not just use rowBuf?

Reply:
> check if the upstream tikv-client-java need to be fixed

OK,I will check

Reply:
> why need mayConflictRowKeys, and not just use rowBuf?

There are two reasons here.
1. the primary keys of inserted rows are not necessarily conflicting, we only know which columns are really conflicting after we send all the primary keys that may conflict to TiKV and get the returned results from TiKV.
2. the primary keys of the columns with conflicting unique indexes may be the same as the primary keys in the inserted columns, which are stored using Set variables for de-duplication.